# Determine Cost Function for the Artificial Neural Network for Iris Train and Test

## Define functions, read training file, and get cost function value

In [1]:
# Import numpy
import numpy as np

# Unity/Identity activation function for layer 0
def identity_activation(G0):
    return G0

# TanH activation function for layer 1
def tanh_activation(G1):
    return (np.exp(G1)-np.exp(-G1))/(np.exp(G1)+np.exp(-G1))

# Sigmoid activation function for layer 2
def sigmoid_activation(G2):
    return 1/(1+np.exp(-G2))

# Initialize Weights and Biases
def initialize_weights_and_biases(n):
    W0 = np.identity(n)
    b0 = np.zeros((n,1))
    
    W1 = np.full((n,n),0.5)
    b1 = np.full((n,1),0.5)
    
    W2 = np.full((1,n),0.5)
    b2 = np.full((1,1),0.5)
    
    return W0, b0, W1, b1, W2, b2

# Forward Propagation
def forward_prop(X0, W0, b0, W1, b1, W2, b2):

    # Layer 0 - Unity
    X1 = identity_activation(W0@X0 + b0)
    
    # Layer 1 - TanH
    X2 = tanh_activation(W1@X1 + b1)
    
    # Layer 2 - Sigmoid
    Yhat = sigmoid_activation(W2@X2 + b2)
    
    return Yhat

# Cost function
def cost_function(Y, Yhat, m):
    return -(1/m)*np.sum(Y*np.log(Yhat) + (1-Y)*np.log(1-Yhat))

# Read CSV data from input file
train_csv = np.genfromtxt ('iristrain-1.csv', delimiter=",")

X_train = train_csv[1:,0:4].transpose()
Y_train = train_csv[1:,4:5].transpose()

# Number of training examples
m_train = X_train.shape[1]

# Number of features
n_train = X_train.shape[0]

# Initialize the Weights and Biases
W0_train, b0_train, W1_train, b1_train, W2_train, b2_train = initialize_weights_and_biases(n_train)

# Forward Propagation through the layers 0, 1 and 2
Yhat_train = forward_prop(X_train, W0_train, b0_train, W1_train, b1_train, W2_train, b2_train)

# Calculate the Cost Function
J_train = cost_function(Y_train, Yhat_train, m_train)

print("The cost function for the training dataset is: {}".format(J_train))

The cost function for the training dataset is: 1.328851528513772


## Read testing file and get cost function value

In [2]:

# Read CSV data from input file
test_csv = np.genfromtxt ('iristest-1.csv', delimiter=",")

X_test = test_csv[1:,0:4].transpose()
Y_test = test_csv[1:,4:5].transpose()

# Number of training examples
m_test = X_test.shape[1]

# Number of features
n_test = X_test.shape[0]

# Initialize the Weights and Biases
W0_test, b0_test, W1_test, b1_test, W2_test, b2_test = initialize_weights_and_biases(n_test)

# Forward Propagation through the layers 0, 1 and 2
Yhat_test = forward_prop(X_test, W0_test, b0_test, W1_test, b1_test, W2_test, b2_test)

# Calculate the Cost Function
J_test = cost_function(Y_test, Yhat_test, m_test)

print("The cost function for the testing dataset is: {}".format(J_test))


The cost function for the testing dataset is: 1.328862773571919


The results should be printed as follows ---

The cost function for the training dataset is: 1.328851528513772

The cost function for the testing dataset is:  1.328862773571919